projects/components/src/viewer/code-viewer/code-viewer.component.ts

```diff
@@ -0,0 +1,127 @@
+import { ChangeDetectionStrategy, Component, Input, TemplateRef } from '@angular/core';
+import { Color } from '@hypertrace/common';
+import { isEmpty } from 'lodash-es';
+import { of } from 'rxjs';
+import { DownloadFileMetadata } from '../../download-file/download-file-metadata';
+
+@Component({
+  selector: 'ht-code-viewer',
+  styleUrls: ['./code-viewer.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  template: `
+    <div *ngIf="this.code" class="code-viewer" [style.backgroundColor]="this.backgroundColor">
+      <div *ngIf="this.showHeader" class="header">
+        <div class="title">{{ this.titleText }}</div>
+        <div class="header-content">
+          <ng-container *ngIf="this.additionalHeaderContent">
+            <ng-container *ngTemplateOutlet="this.additionalHeaderContent"></ng-container
+          ></ng-container>
+          <ht-search-box
+            *ngIf="this.enableSearch"
+            class="search-box"
+            (valueChange)="this.onSearch($event)"
+          ></ht-search-box>
+          <ht-download-file
+            *ngIf="!!this.code && !!this.downloadFileName"
+            [metadata]="this.downloadCodeMetadata"
+          ></ht-download-file>
+        </div>
+      </div>
+      <div class="content">
+        <div class="line-numbers">
+          <div
+            *ngFor="let lineNumber of this.lineNumbers"
```

Comment:
even though `lineNumbers`, codeLines are getters, they would still be executed with every `do check` cycle.  if we have a huge spec then this could be a costly operation. We could just calculate it one time inside that ngOnChanges. Any reason for not doing that and choosing getter approach?

Reply:
same for `downloadCodeMetadata`

Reply:
done